File: service/src/test/java/hu/bsstudio/robonaut/team/DefaultTeamServiceTest.java
package hu.bsstudio.robonaut.team;

import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

import hu.bsstudio.robonaut.entity.TeamEntity;
import hu.bsstudio.robonaut.entity.TeamType;
import hu.bsstudio.robonaut.repository.TeamRepository;
import hu.bsstudio.robonaut.team.mapper.TeamModelEntityMapper;
import hu.bsstudio.robonaut.team.model.DetailedTeam;
import hu.bsstudio.robonaut.team.model.Team;
import java.util.Collections;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

final class DefaultTeamServiceTest {

    private static final long TEAM_ID_1 = 42;
    private static final long TEAM_ID_2 = 43;
    private static final List<String> TEAM_MEMBERS_1 = List.of("Bence", "Boldi");
    private static final List<String> TEAM_MEMBERS_2 = List.of("Bence", "Boldi", "Máté");
    private static final String TEAM_NAME_1 = "BSS";
    private static final String TEAM_NAME_2 = "Budavári Schönherz Stúdió";
    private static final TeamType TEAM_TYPE_1 = TeamType.SENIOR;
    private static final TeamType TEAM_TYPE_2 = TeamType.JUNIOR;
    private static final int YEAR_1 = 2020;
    private static final int YEAR_2 = 2021;
    public static final TeamEntity TEAM_ENTITY_1 = createTeamEntity(TEAM_ID_1, YEAR_1, TEAM_NAME_1, TEAM_MEMBERS_1, TEAM_TYPE_1);
    public static final TeamEntity OLD_TEAM_ENTITY = createTeamEntity(TEAM_ID_1, YEAR_2, TEAM_NAME_2, TEAM_MEMBERS_2, TEAM_TYPE_2);
    private static final Team TEAM_1 = Team.builder()
        .teamId(TEAM_ID_1)
        .teamMembers(TEAM_MEMBERS_1)
        .teamName(TEAM_NAME_1)
        .teamType(TEAM_TYPE_1)
        .year(YEAR_1)
        .build();

    private DefaultTeamService underTest;

    @Mock
    private TeamRepository mockRepository;
    @Mock
    private TeamModelEntityMapper mockMapper;

    @BeforeEach
    void setUp() {
        initMocks(this);
        this.underTest = new DefaultTeamService(mockRepository);
        this.underTest.setTeamMapper(mockMapper);
    }

    @Test
    void shouldReturnCreatedTeam() {
        when(mockRepository.insert(TEAM_ENTITY_1)).thenReturn(Mono.just(TEAM_ENTITY_1));
        final var detailedTeam = DetailedTeam.builder().build();
        when(mockMapper.toModel(TEAM_ENTITY_1)).thenReturn(detailedTeam);

        final var result = underTest.addTeam(TEAM_1);

        StepVerifier.create(result)
            .expectNext(detailedTeam)
            .verifyComplete();
    }

    @Test
    void shouldReturnUpdatedTeam() {
        when(mockRepository.findById(TEAM_ID_1)).thenReturn(Mono.just(OLD_TEAM_ENTITY));
        final var updatedTeamEntity = TEAM_ENTITY_1;
        when(mockRepository.save(updatedTeamEntity)).thenReturn(Mono.just(updatedTeamEntity));
        final var detailedTeam = DetailedTeam.builder().build();
        when(mockMapper.toModel(updatedTeamEntity)).thenReturn(detailedTeam);

        final var result = underTest.updateTeam(TEAM_1);

        StepVerifier.create(result)
            .expectNext(detailedTeam)
            .verifyComplete();
    }

    @Test
    void shouldReturnAllTeam() {
        final var teamEntity1 = new TeamEntity();
        teamEntity1.setTeamId(TEAM_ID_1);
        final var teamEntity2 = new TeamEntity();
        teamEntity2.setTeamId(TEAM_ID_2);
        when(mockRepository.findAll()).thenReturn(Flux.just(teamEntity1, teamEntity2));
        final var detailedTeam1 = DetailedTeam.builder().teamId(TEAM_ID_1).build();
        final var detailedTeam2 = DetailedTeam.builder().teamId(TEAM_ID_2).build();
        when(mockMapper.toModel(teamEntity1)).thenReturn(detailedTeam1);
        when(mockMapper.toModel(teamEntity2)).thenReturn(detailedTeam2);

        final var result = underTest.findAllTeam();

        StepVerifier.create(result)
            .expectNext(detailedTeam1)
            .expectNext(detailedTeam2)
            .verifyComplete();
    }

    private static TeamEntity createTeamEntity(final Long teamId, final int year, final String teamName,
                                               final List<String> teamMembers, final TeamType teamType) {
        final var teamEntity = new TeamEntity();
        teamEntity.setTeamId(teamId);
        teamEntity.setYear(year);
        teamEntity.setTeamName(teamName);
        teamEntity.setTeamMembers(teamMembers);
        teamEntity.setTeamType(teamType);
        teamEntity.setSpeedTimes(Collections.emptyList());
        return teamEntity;
    }
}

File: service/src/test/kotlin/hu/bsstudio/robonaut/team/DefaultTeamServiceTest.kt
package hu.bsstudio.robonaut.team

import hu.bsstudio.robonaut.entity.ScoreEntity
import hu.bsstudio.robonaut.entity.TeamEntity
import hu.bsstudio.robonaut.entity.TeamType
import hu.bsstudio.robonaut.repository.TeamRepository
import hu.bsstudio.robonaut.team.mapper.TeamModelEntityMapper
import hu.bsstudio.robonaut.team.model.DetailedTeam
import hu.bsstudio.robonaut.team.model.Score
import hu.bsstudio.robonaut.team.model.Team
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.test.StepVerifier

internal class DefaultTeamServiceTest {
    @MockK
    private lateinit var mockRepository: TeamRepository

    @MockK
    private lateinit var mockMapper: TeamModelEntityMapper

    private lateinit var underTest: DefaultTeamService

    @BeforeEach
    internal fun setUp() {
        MockKAnnotations.init(this)
        underTest = DefaultTeamService(mockRepository)
        underTest = DefaultTeamService(mockRepository, mockMapper)
    }

    @Test
    internal fun `should return created Team`() {
        every { mockRepository.insert(TEAM_ENTITY_1) } returns Mono.just(TEAM_ENTITY_1)
        every { mockMapper.toModel(TEAM_ENTITY_1) } returns DETAILED_TEAM_1

        val result = underTest.addTeam(TEAM_1)

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete()
    }

    @Test
    internal fun `should return updated Team`() {
        every { mockRepository.findById(TEAM_ID_1) } returns Mono.just(OLD_TEAM_ENTITY)
        every { mockRepository.save(TEAM_ENTITY_1) } returns Mono.just(TEAM_ENTITY_1)
        every { mockMapper.toModel(TEAM_ENTITY_1) } returns DETAILED_TEAM_1

        val result = underTest.updateTeam(TEAM_1)

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete()
    }

    @Test
    internal fun `should return updated Team by admin`() {
        every { mockMapper.toEntity(DETAILED_TEAM_1) } returns TEAM_ENTITY_1
        every { mockRepository.save(TEAM_ENTITY_1) } returns Mono.just(TEAM_ENTITY_1)
        every { mockMapper.toModel(TEAM_ENTITY_1) } returns DETAILED_TEAM_1

        val result = underTest.updateTeam(DETAILED_TEAM_1)

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete()
    }

    @Test
    internal fun `should return all Team`() {
        every { mockRepository.findAll() } returns Flux.just(TEAM_ENTITY_1, TEAM_ENTITY_2)
        every { mockMapper.toModel(TEAM_ENTITY_1) } returns DETAILED_TEAM_1
        every { mockMapper.toModel(TEAM_ENTITY_2) } returns DETAILED_TEAM_2

        val result = underTest.findAllTeam()

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .expectNext(DETAILED_TEAM_2)
            .verifyComplete()
    }

    companion object {
        private const val TEAM_ID_1: Long = 42
        private const val TEAM_ID_2: Long = 43
        private val TEAM_MEMBERS_1 = listOf("Bence", "Boldi")
        private val TEAM_MEMBERS_2 = listOf("Bence", "Boldi", "Máté")
        private const val TEAM_NAME_1 = "BSS"
        private const val TEAM_NAME_2 = "Budavári Schönherz Stúdió"
        private val TEAM_TYPE_1 = TeamType.SENIOR
        private val TEAM_TYPE_2 = TeamType.JUNIOR
        private const val YEAR_1 = 2020
        private const val YEAR_2 = 2021
        private val OLD_TEAM_ENTITY = createTeamEntity(TEAM_ID_1, YEAR_2, TEAM_NAME_2, TEAM_MEMBERS_2, TEAM_TYPE_2)
        private val TEAM_ENTITY_1 = createTeamEntity(TEAM_ID_1, YEAR_1, TEAM_NAME_1, TEAM_MEMBERS_1, TEAM_TYPE_1)
        private val TEAM_ENTITY_2 = createTeamEntity(TEAM_ID_2, YEAR_2, TEAM_NAME_2, TEAM_MEMBERS_2, TEAM_TYPE_2)
        private val TEAM_1: Team =
            Team(
                teamId = TEAM_ID_1,
                teamMembers = TEAM_MEMBERS_1,
                teamName = TEAM_NAME_1,
                teamType = TEAM_TYPE_1,
                year = YEAR_1,
            )
        private val DETAILED_TEAM_1: DetailedTeam =
            DetailedTeam(
                teamId = TEAM_ID_1,
                combinedScore = Score(0, 0, 0),
                juniorScore = Score(0, 0, 0),
            )
        private val DETAILED_TEAM_2 = DetailedTeam(teamId = TEAM_ID_2)

        private fun createTeamEntity(
            teamId: Long,
            year: Int,
            teamName: String,
            teamMembers: List<String>,
            teamType: TeamType,
        ): TeamEntity {
            val defaultScore = ScoreEntity()
            return TeamEntity(
                teamId = teamId,
                year = year,
                teamName = teamName,
                teamMembers = teamMembers,
                teamType = teamType,
                speedTimes = emptyList(),
                score = defaultScore,
                juniorScore = defaultScore,
            )
        }
    }
}
